packages/graphql/src/schema/validation/custom-rules/directive-argument-of-correct-type.ts

```diff
@@ -22,67 +22,83 @@ import type {
     DirectiveNode,
     GraphQLArgument,
     ArgumentNode,
-    ASTNode,
-    ObjectTypeDefinitionNode,
-    FieldDefinitionNode,
+    GraphQLDirective,
+    GraphQLSchema,
 } from "graphql";
-import { GraphQLError, coerceInputValue, valueFromASTUntyped, buildASTSchema } from "graphql";
-
+import { coerceInputValue, valueFromASTUntyped, buildASTSchema } from "graphql";
+import type { Maybe } from "graphql/jsutils/Maybe";
 import type { SDLValidationContext } from "graphql/validation/ValidationContext";
 import { VALIDATION_ERROR_CODES } from "../utils/validation-error-codes";
+import type { AssertionResponse } from "./utils/document-validation-error";
+import { createGraphQLError } from "./utils/document-validation-error";
+import { getPathToNode } from "./utils/path-parser";
 
 export function DirectiveArgumentOfCorrectType(context: SDLValidationContext): ASTVisitor {
-    const schema = buildASTSchema(context.getDocument(), { assumeValid: true, assumeValidSDL: true });
+    // TODO: find a way to scope this schema instead of creating the whole document
+    // should only contain dynamic directives and their associated types (typeWhere + jwt payload)
+    let schema: GraphQLSchema | undefined;
+    const getSchemaFromDocument = (): GraphQLSchema => {
+        if (!schema) {
+            schema = buildASTSchema(context.getDocument(), { assumeValid: true, assumeValidSDL: true });
+        }
+        return schema;
+    };
 
     return {
         Directive(directiveNode: DirectiveNode, _key, _parent, path, ancenstors) {
-            const genericDirectiveName = ["subscriptionsAuthorization", "authorization", "authentication"].find(
+            const oneOfAuthorizationDirectives = [
+                "subscriptionsAuthorization",
+                "authorization",
+                "authentication",
+            ].reduce<string | undefined>((genericDirective, oneOfAuthorizationDirectives) => {
+                if (directiveNode.name.value.toLowerCase().includes(oneOfAuthorizationDirectives)) {
+                    genericDirective = oneOfAuthorizationDirectives;
+                }
+                return genericDirective;
+            }, undefined);
+            const otherDirectives = ["fulltext", "relationship", "node", "customResolver", "cypher"].find(
                 (applicableDirectiveName) =>
-                    directiveNode.name.value.toLowerCase().includes(applicableDirectiveName.toLowerCase())
+                    directiveNode.name.value.toLowerCase() === applicableDirectiveName.toLowerCase()
             );
-            // Validate only Authorization/Authentication usage
-            if (!genericDirectiveName) {
+
+            if (!oneOfAuthorizationDirectives && !otherDirectives) {
                 return;
             }
-            const directiveDefinition = schema.getDirective(directiveNode.name.value);
+
+            let directiveName: string | undefined;
```

Comment:
this is for the case of authorization directives, because for example:
```js
type User @authorization(...) {
...
}
``` 
the directiveNode.name.value is going to be "authorization" while the actual directiveName is going to be "UserAuthorization"